### ISSUE TITLE
Understanding CAD OSE

### ISSUE THREAD

**Issue by NoamGit** [numenta/NAB#336]
Hi,

I am trying to understand the CAD-OSE, however since @smirmik lacks proper documentation, I find it to be a tough and very frustrating task.
I am familiar with the ideas of HTM, and I am trying to build the right analogies for the code. Especially, I find the memory storage structure, context, and facts unintuitive.
I would like to understand the following:
- What is the right and left parts of a `Context`, and what exactly is `semiContext`s (left and right?)
- What is the definition of `Facts`?
And in general, what is the purpose of the functions:
- `step`
- `contextCrosser`

Can anyone shed some lite on the simple but powerful algorithm?

Thanks

**Comment by subutai** [comment]
Good questions. Hopefully @smirmik will respond. It would be good to have a better understanding of this (externally contributed) algorithm.

**Comment by smirmik** [comment]
Hello @NoamGit ,

Thank you for your interest!

I have some health problems with my spine and I can not sit at the computer for a long time. I am on treatment. However, your panic in all repositories and in e-mail forces me.

CAD Based on my Contextual Memory (CM) project. CM was developed to create a General Artificial Intelligence model. I was looking for a solution for more than 18 years and only in 2016 I came up with the first version of CM. Based on CM, I made a CAD and it showed itself perfectly in the NAB competition. I was not inspired by some existing development (which is why CM was created). This is completely my development. But I do not exclude that there is something similar in the world. I did not write articles and a normal description of the algorithm does not exist at the moment. 
There is only one incomplete description that I send to those who email me. It does not fully correspond to the latest version of the code, but it can help to understand the basic idea.

Currently i have another version of CAD. This is about 15-20 times faster than HTM with comparable results on one-dimensional data. There is also a version that shows better results than HTM, with comparable speed.
CAD optimization continues, and in the future it will be even faster.

Maybe I will post the changes in the python code and post them to this repository later if the guys from Numenta are not opposed.
